File: 17.电话号码的字母组合.go


/*
 * @lc app=leetcode.cn id=17 lang=golang
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start

func backtrack(curStr string, res *[]string, index int, digits string, digitsMap map[byte][]string) {
	if index == len(digits) {
		*res = append(*res, curStr)
		return
	}
	for _, s := range digitsMap[digits[index]] {
		backtrack(curStr+s, res, index+1, digits, digitsMap)
	}
}

func letterCombinations(digits string) []string {
	var dict = map[byte][]string{
		'0': {"0"},
		'1': {"1"},
		'2': {"a", "b", "c"},
		'3': {"d", "e", "f"},
		'4': {"g", "h", "i"},
		'5': {"j", "k", "l"},
		'6': {"m", "n", "o"},
		'7': {"p", "q", "r", "s"},
		'8': {"t", "u", "v"},
		'9': {"w", "x", "y", "z"},
	}

	var res = []string{}

	if len(digits) == 0 {
		return res
	}

	backtrack("", &res, 0, digits, dict)

	return res
}

// @lc code=end

// func main() {
// 	letterCombinations("234")
// }

File: leetcode_ts/src/17.电话号码的字母组合.ts
/*
 * @lc app=leetcode.cn id=17 lang=typescript
 *
 * [17] 电话号码的字母组合
 */

// @lc code=start

function backtrack(curStr: string, res: string[], index: number, digits: string, dict: Record<string, string[]>) {
  if(index === digits.length){
    res.push(curStr);
  }

  for(let i in dict[digits[index]]){
    const ch = dict[digits[index]][i];
    backtrack(curStr+ch,res,index+1,digits,dict);
  }

}

function letterCombinations(digits: string): string[] {
  const dict: Record<string, string[]> = {
    '0': ["0"],
    '1': ["1"],
    '2': ["a", "b", "c"],
    '3': ["d", "e", "f"],
    '4': ["g", "h", "i"],
    '5': ["j", "k", "l"],
    '6': ["m", "n", "o"],
    '7': ["p", "q", "r", "s"],
    '8': ["t", "u", "v"],
    '9': ["w", "x", "y", "z"]
  }

  const res: string[] = [];

  if (digits === "") {
    return res;
  }

  backtrack("",res,0,digits,dict);

  return res;
};
// @lc code=end

letterCombinations("234");